File: src/singler/lib/train_single_reference.cpp
#include "utils.h" // must be before all other includes.

#include <vector>
#include <cstdint>
#include <algorithm>

//[[export]]
void* train_single_reference(void* ref, const int32_t* labels /** numpy */, void* markers, uint8_t approximate, int32_t nthreads) {
    singlepp::BasicBuilder builder;
    builder.set_num_threads(nthreads);
    builder.set_top(-1); // Use all available markers; assume subsetting was applied on the Python side.
    builder.set_approximate(approximate);

    auto marker_ptr = reinterpret_cast<const singlepp::Markers*>(markers);
    const auto& ptr = reinterpret_casts<const tatami::Mattress*>(ref)->ptr;
    std::vector<int> labels2(labels, labels + ptr->ncol()); // need to copy as int may not be int32 and singlepp isn't templated on the labels (for now).
    auto built = builder.run(ptr.get(), labels2.data(), *marker_ptr);

    return new singlepp::BasicBuilder::Prebuilt(std::move(built));
}

//[[export]]
int32_t get_nsubset_from_single_reference(void* ptr) {
    return reinterpret_cast<const singlpp::BasicBuilder::Prebuilt*>(ptr)->subset.size();
}

//[[export]]
int32_t get_nlabels_from_single_reference(void* ptr) {
    return reinterpret_cast<const singlpp::BasicBuilder::Prebuilt*>(ptr)->num_labels();
}

//[[export]]
void get_subset_from_single_reference(void* ptr, int32_t* buffer /** numpy */) {
    const auto& sub = reinterpret_cast<const singlpp::BasicBuilder::Prebuilt*>(ptr)->subset;
    std::copy(sub.begin(), sub.end(), buffer);
}

//[[export]]
void free_single_reference(void* ptr) {
    delete reinterpret_cast<singlepp::BasicBuilder::Prebuilt>(pr);
}

File: src/singler/train_single_reference.py
from .InternalMarkers import InternalMarkers
from . import cpphelpers as lib
from .utils import _factorize
from mattress import tatamize
from numpy import int32


class SinglePrebuiltReference:
    def __init__(self, ptr):
        self._ptr = ptr

    def __del__(self):
        lib.free_prebuilt(self._ptr)

    def num_features(self):
        return lib.get_nsubset_from_prebuilt(self._ptr)

    def num_labels(self):
        return lib.get_nlabels_from_prebuilt(self._ptr)

    def subset(self, features: Sequence):
        nmarkers = self.num_features()
        buffer = ndarray(nmarkers, dtype=int32)
        lib.get_subset_from_prebuilt(self._ptr, buffer);
        return [features[i] for i in buffer]


def train_single_reference(
    ref,
    labels: Sequence,
    features: Sequence,
    markers: dict[Any, dict[Any, Sequence]],
    approximate: bool = True,
    num_threads: int = 1,
):
    lablev, labind = _factorize(labels)
    mrk = InternalMarkers.from_dict(markers, lablev, features)
    mat_ptr = tatamize(ref)

    return SinglePrebuiltReference(
        lib.prebuild_reference(
            mat_ptr.ptr, 
            labels = labind, 
            markers = mrk.ptr, 
            approximate = approximate, 
            num_threads = num_threads
        )
    )
